relay-aws-extension/src/aws_extension.rs

```diff
@@ -0,0 +1,193 @@
+use reqwest::StatusCode;
+use serde::Deserialize;
+use std::collections::HashMap;
+
+const EXTENSION_NAME: &str = "sentry-lambda-extension";
+const EXTENSION_NAME_HEADER: &str = "Lambda-Extension-Name";
+const EXTENSION_ID_HEADER: &str = "Lambda-Extension-Identifier";
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+struct RegisterResponse {
+    pub function_name: String,
+    pub function_version: String,
+    pub handler: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+struct Tracing {
+    pub r#type: String,
+    pub value: String,
+}
+
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)]
+#[serde(rename_all = "UPPERCASE", tag = "eventType")]
+enum NextEventResponse {
+    #[serde(rename_all = "camelCase")]
+    Invoke {
+        deadline_ms: u64,
+        request_id: String,
+        invoked_function_arn: String,
+        tracing: Tracing,
+    },
+    #[serde(rename_all = "camelCase")]
+    Shutdown {
+        shutdown_reason: String,
+        deadline_ms: u64,
+    },
+}
+
+#[derive(Debug, Fail)]
+#[fail(display = "aws extension error")]
+/// Generic error in an AWS extension context
+pub struct AwsExtensionError;
+
+/// Actor implementing an AWS extension
+///
+/// Spawns an actor that
+/// * registers with the AWS extensions API
+/// * sends blocking NextEvent calls to the extensions API to get the next invocation
+///
+/// Each finished invocation immediately polls for the next event. Note that AWS might
+/// freeze the container indefinitely for unused lambdas and this request will also wait
+/// till things are active again.
+pub struct AwsExtension {
+    /// The base url for the AWS Extensions API
+    base_url: String,
+    /// The reqwest client needs to be blocking and with 0 timeout
+    /// because the container might get frozen if lambda is unused
+    reqwest_client: Client,
+    /// The extension id that will be retrieved on register and used
+    /// for subsequent requests
+    extension_id: Option<String>,
+}
+
+impl AwsExtension {
+    /// Creates a new `AwsExtension` instance.
+    pub fn new(aws_runtime_api: String) -> Result<Self, AwsExtensionError> {
+        let base_url = format!("http://{}/2020-01-01/extension", aws_runtime_api);
+
+        let reqwest_client = reqwest::blocking::Client::builder()
+            .timeout(None)
+            .build()
+            .map_err(|_| AwsExtensionError)?;
+
+        Ok(AwsExtension {
+            base_url,
+            reqwest_client,
+            extension_id: None,
+        })
+    }
+
+    fn register(&mut self) -> Result<(), AwsExtensionError> {
+        let url = format!("{}/register", self.base_url);
+        let map = HashMap::from([("events", ["INVOKE", "SHUTDOWN"])]);
+
+        let res = self
+            .reqwest_client
+            .post(&url)
+            .header(EXTENSION_NAME_HEADER, EXTENSION_NAME)
+            .json(&map)
+            .send()
```

Comment:
You're using the blocking client here, which is not ideal since it blocks the arbiter this actor is running in. You have two options:
1. Use the non-blocking client in a tokio 1 runtime. We will soon clean this all up to make it easier, but in the meanwhile, you'll want to implement it like here (which means, you can use `.await` to chain several calls): https://github.com/getsentry/relay/blob/ef952cb5b6e41c1dd41c4f4b2e268e28741063ca/relay-server/src/actors/upstream.rs#L581-L587
2. Document that this actor _must_ run in its completely own thread and place a big TODO on top of this to clean it up as soon as we've migrated to tokio 1. However, that also means you _cannot_ receive messages such as the shutdown while you're waiting for a response, so I'm not sure if that's even a feasible option.

Given the similarity to `UpstreamRelay`, I'd prefer option 1.

Reply:
I would also prefer option 1. Note that with option 1, the entire actor can be one big async block containing a `loop {}` (no need for context.notify)

Reply:
It may be useful to keep the outermost loop in a `notify` cycle so we can break it more easily from the outside, although that doesn't seem required for AWS environments.

Reply:
this is done now @jan-auer @untitaker, if you guys can take a look that the approach is ok?